File: 0_iterator-patern_ts/src/BookShelf.ts
import Aggregate from 'Aggregate';
import Book from 'Book';
import Iterator from 'Iterator';
import BookShelfIterator from 'BookShelfIterator';

export default class BookShelf implements Aggregate {
    private books: Book[] = new Array<Book>();

    constructor() {}

    public getBookAt(index: number): Book {
        return this.books[index];
    }

    public appendBook(book: Book): void {
        this.books.push(book);
    }

    public getLength(): number {
        return this.books.length;
    }

    public iterator(): BookShelfIterator {
        return new BookShelfIterator(this);
    }
}

File: 0_iterator-patern_hw/BookShelf.java
import java.util.ArrayList;

public class BookShelf implements Aggregate {
    private ArrayList<Book> books = new ArrayList<Book>();

    public BookShelf() {}

    public Book getBookAt(int index) {
        return books.get(index);
    }

    public void appendBook(Book book) {
        this.books.add(book);
    }

    public int getLength() {
        return this.books.size();
    }

    public Iterator iterator() {
        return new BookShelfIterator(this);
    }
}